src/zenml/utils/metadata_utils.py

```diff
@@ -0,0 +1,339 @@
+    metadata: Dict[str, MetadataType],
+    run_id_name_or_prefix: Union[UUID, str],
+) -> None: ...
+
+
+@overload
+def log_metadata(
+    *,
+    metadata: Dict[str, MetadataType],
+    step_name: str,
+    run_id_name_or_prefix: Union[UUID, str],
+) -> None: ...
+
+
+def log_metadata(
+    metadata: Dict[str, MetadataType],
+    # Parameters to manually log metadata for steps and runs
+    step_id: Optional[UUID] = None,
+    step_name: Optional[str] = None,
+    run_id_name_or_prefix: Optional[Union[UUID, str]] = None,
+    # Parameters to manually log metadata for artifacts
+    artifact_version_id: Optional[UUID] = None,
+    artifact_name: Optional[str] = None,
+    artifact_version: Optional[str] = None,
+    # Parameters to manually log metadata for models
+    model_version_id: Optional[UUID] = None,
+    model_name: Optional[str] = None,
+    model_version: Optional[str] = None,
+) -> None:
+    """Logs metadata for various resource types in a generalized way.
+
+    Args:
+        metadata: The metadata to log.
+        step_id: The ID of the step.
+        step_name: The name of the step.
+        run_id_name_or_prefix: The id, name or prefix of the run
+        artifact_version_id: The ID of the artifact version
+        artifact_name: The name of the artifact.
+        artifact_version: The version of the artifact.
+        model_version_id: The ID of the model version.
+        model_name: The name of the model.
+        model_version: The version of the model
+
+    Raises:
+        ValueError: If no identifiers are provided and the function is not
+            called from within a step.
+    """
+    # Initialize the client
+    client = Client()
+
+    # If a step name is provided, we need a run_id_name_or_prefix and will log
+    # metadata for the steps pipeline and model accordingly.
+    if step_name is not None and run_id_name_or_prefix is not None:
+        run_model = client.get_pipeline_run(
+            name_id_or_prefix=run_id_name_or_prefix
+        )
+        step_model = run_model.steps[step_name]
+
+        client.create_run_metadata(
+            metadata=metadata,
+            resource_id=run_model.id,
+            resource_type=MetadataResourceTypes.PIPELINE_RUN,
+        )
+        client.create_run_metadata(
+            metadata=metadata,
+            resource_id=step_model.id,
+            resource_type=MetadataResourceTypes.STEP_RUN,
+        )
+        if step_model.model_version:
+            client.create_run_metadata(
+                metadata=metadata,
+                resource_id=step_model.model_version.id,
+                resource_type=MetadataResourceTypes.MODEL_VERSION,
+            )
+
+    # If a step is identified by id, fetch it directly through the client,
+    # follow a similar procedure and log metadata for its pipeline and model
+    # as well.
+    elif step_id is not None:
+        step_model = client.get_run_step(step_run_id=step_id)
+        run_model = client.get_pipeline_run(
```

Comment:
This is just an unnecessary server call, why fetch the pipeline run if you only need its ID that you already have?

Reply:
Removed the redundant call.